source/extensions/filters/network/postgres_proxy/postgres_decoder.cc

```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
   };
 }
 
-Decoder::Result DecoderImpl::parseHeader(Buffer::Instance& data) {
-  ENVOY_LOG(trace, "postgres_proxy: parsing message, len {}", data.length());
+/* Main handler for incoming messages. Messages are dispatched based on the
+   current decoder's state.
+*/
+Decoder::Result DecoderImpl::onData(Buffer::Instance& data, bool frontend) {
+  switch (state_) {
+  case State::InitState:
+    return onDataInit(data, frontend);
+  case State::OutOfSyncState:
+  case State::EncryptedState:
+    return onDataIgnore(data, frontend);
+  case State::InSyncState:
+    return onDataInSync(data, frontend);
+  default:
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  }
+}
+
+/* Handler for messages when decoder is in Init State. There are very few message types which
+   are allowed in this state.
+   If the initial message has the correct syntax and  indicates that session should be in
+   clear-text, the decoder will move to InSyncState. If the initial message has the correct syntax
+   and indicates that session should be encrypted, the decoder stays in InitState, because the
+   initial message will be received again after transport socket negotiates SSL. If the message
+   syntax is incorrect, the decoder will move to OutOfSyncState, in which messages are not parsed.
+*/
+Decoder::Result DecoderImpl::onDataInit(Buffer::Instance& data, bool) {
+  ASSERT(state_ == State::InitState);
 
   // The minimum size of the message sufficient for parsing is 5 bytes.
-  if (data.length() < 5) {
+  if (data.length() < 4) {
     // not enough data in the buffer.
-    return Decoder::NeedMoreData;
+    return Decoder::Result::NeedMoreData;
+  }
+
+  // Validate the message before processing.
+  const MsgBodyReader& f = std::get<1>(first_);
+  const auto msgParser = f();
+  // Run the validation.
+  message_len_ = data.peekBEInt<uint32_t>(0);
+// MAX_STARTUP_PACKET_LENGTH is defined in Postgres source code
+// as maximum size of initial packet.
+#define MAX_STARTUP_PACKET_LENGTH 10000
```

Comment:
Perhaps move it to `postgres_decoder.h` and add the following link to comments: https://github.com/postgres/postgres/search?q=MAX_STARTUP_PACKET_LENGTH&type=code